Objects/listobject.c

```diff
@@ -218,7 +218,20 @@ valid_index(Py_ssize_t i, Py_ssize_t limit)
     return (size_t) i < (size_t) limit;
 }
 
-static PyObject *indexerr = NULL;
+static inline void
+index_error(const char *msg, Py_ssize_t len, Py_ssize_t i)
+{
+    if (len == 0) {
+        PyErr_Format(PyExc_IndexError, "%s index out of range, got index %zd but the list is empty", msg, i);
+        return;
+    }
+    if (i < 0)
+        i -= len;
```

Comment:
formatting: PEP 7 needs brackets on new code
```suggestion
    if (i < 0) {
        i -= len;
    }
```